spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/convert/SqlGenerator.java

```diff
@@ -507,45 +508,74 @@ private String createFindAllSql() {
 	}
 
 	private SelectBuilder.SelectWhere selectBuilder() {
-		return selectBuilder(Collections.emptyList());
+		return selectBuilder(Collections.emptyList(), Query.empty());
+	}
+
+	private SelectBuilder.SelectWhere selectBuilder(Query query) {
+		return selectBuilder(Collections.emptyList(), query);
 	}
 
 	private SelectBuilder.SelectWhere selectBuilder(Collection<SqlIdentifier> keyColumns) {
+		return selectBuilder(keyColumns, Query.empty());
+	}
+
+	private SelectBuilder.SelectWhere selectBuilder(Collection<SqlIdentifier> keyColumns, Query query) {
 
 		Table table = getTable();
 
-		Set<Expression> columnExpressions = new LinkedHashSet<>();
+		Projection projection = getProjection(keyColumns, query, table);
+		SelectBuilder.SelectAndFrom selectBuilder = StatementBuilder.select(projection.columns());
+		SelectBuilder.SelectJoin baseSelect = selectBuilder.from(table);
 
-		List<Join> joinTables = new ArrayList<>();
-		for (PersistentPropertyPath<RelationalPersistentProperty> path : mappingContext
-				.findPersistentPropertyPaths(entity.getType(), p -> true)) {
+		for (Join join : projection.joins()) {
+			baseSelect = baseSelect.leftOuterJoin(join.joinTable).on(join.joinColumn).equals(join.parentId);
+		}
 
-			AggregatePath extPath = mappingContext.getAggregatePath(path);
+		return (SelectBuilder.SelectWhere) baseSelect;
+	}
 
-			// add a join if necessary
-			Join join = getJoin(extPath);
-			if (join != null) {
-				joinTables.add(join);
+	private Projection getProjection(Collection<SqlIdentifier> keyColumns, Query query, Table table) {
+
+		Set<Expression> columns = new LinkedHashSet<>();
+		Set<Join> joins = new LinkedHashSet<>();
+
+		if (!CollectionUtils.isEmpty(query.getColumns())) {
+			for (SqlIdentifier columnName : query.getColumns()) {
+				columns.add(Column.create(columnName, table));
```

Comment:
Column selection is missing the mapping from property name to column name. Columns (property paths specifically) should be mapped. Ideally, we can reuse the loop below regarding its functionality with just the input columns to derive property paths from. Also, we do not assume that all columns are mapped onto our domain entities.

Reply:
Done.